### ISSUE TITLE
Different Gradient for Distributed Data Parallel Training

### ISSUE THREAD

**Issue by XuanZhangg** [UKPLab/sentence-transformers#3415]
## 🚩 Gradient Mismatch Between DDP and Single-GPU for `SentenceTransformerTrainer`

**Describe the bug**  
I found that when training a `SentenceTransformer` using `SentenceTransformerTrainer`, the gradients are **different** between single-GPU and multi-GPU (DDP) runs, even with exactly the same code, seed, and data.

**Reproducible Example**

```python
def main():
    from sentence_transformers import SentenceTransformer, SentenceTransformerTrainer, losses
    from sentence_transformers.training_args import SentenceTransformerTrainingArguments
    from datasets import Dataset

    training_args = SentenceTransformerTrainingArguments(
        num_train_epochs=3,
        per_device_train_batch_size=1,
        save_strategy='no'
    )
    model = SentenceTransformer("microsoft/mpnet-base")
    train_dataset = Dataset.from_dict({
        "sentence1": ["It's nice weather outside today.", "He drove to work."],
        "sentence2": ["It's so sunny.", "She walked to the store."],
        "score": [1.0, 0.3],
    })
    loss = losses.CosineSimilarityLoss(model)

    trainer = SentenceTransformerTrainer(
        model=model,
        train_dataset=train_dataset,
        loss=loss,
        args=training_args
    )
    trainer.train()

if __name__ == "__main__":
    main()
```

I added this debug print **inside the Trainer loop before accelerator.backward**, which is in the training_step function:

```python
rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
for name, param in self.model.named_parameters():
    if param.grad is not None:
        print(
            f"Rank: {rank} | Param: {name} | Device: {param.device} | "
            f"Grad norm: {param.grad.norm().item()} | Grad:{param.grad} | param :{param} \n"
        )
    else:
        print(
            f"Rank: {rank} | Param: {name} | Device: {param.device} | Grad: None \n"
        )
    break  # Just print the first param for illustration
```

**Observed output**

✅ **Single GPU:**  
```
Rank: 0 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:0 | Grad norm: 0.429... | Grad: [[ 0.0011, 0.0080, ...]]
```

✅ **DDP (2 GPUs):**  
```
Rank: 0 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:0 | Grad norm: 0.476... | Grad: [[-0.0053, -0.0001, ...]]
Rank: 1 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:1 | Grad norm: 0.476... | Grad: [[-0.0053, -0.0001, ...]]
```

As you can see, the gradients are different — not just scaled, but with different signs and values.

**Expected behavior**  
The gradients should match across single-GPU and DDP runs (up to numerical tolerance) if the setup, data, and loss are the same.

**Comment by tomaarsen** [comment]
Hello!

Apologies in advance if I'm wrong, but if you're using `per_device_train_batch_size=1`, then with single-GPU this means that you have a global/total batch size of 1, whereas with DDP (2 GPUs) you have a global batch size of 2. This means that more data is consumed, different loss, etc.
We wouldn't expect identical gradients and grad norms in that case.
Perhaps a test of `per_device_train_batch_size=1` in the DDP setting and `per_device_train_batch_size=2` in the single-GPU setting should be equivalent, but I'm not sure if it should be.

- Tom Aarsen

**Comment by XuanZhangg** [comment]
Hi Tom, thank you for reply! What I did is to use batchsize=2 for single gpu and 1 for DDP. Basically, what I did to run the code is:

single gpu:
`
export CUDA_VISIBLE_DEVICES=0
`
`
python debug.py --per_device_train_batch_size 2
`
2 GPU:
`torchrun --nproc_per_node=2 debug.py --per_device_train_batch_size 1`

**Comment by Francovelizz** [comment]
> ## 🚩 Desajuste de gradiente entre DDP y una sola GPU para`SentenceTransformerTrainer`
> **Describe el error** que encontré: al entrenar un `SentenceTransformer`using `SentenceTransformerTrainer`, los gradientes son **diferentes** entre ejecuciones de una sola GPU y de múltiples GPU (DDP), incluso con exactamente el mismo código, semilla y datos.
> 
> **Ejemplo reproducible**
> 
> def main():
>     import argparse
>     from sentence_transformers import SentenceTransformer, SentenceTransformerTrainer, losses
>     from sentence_transformers.training_args import SentenceTransformerTrainingArguments
>     from datasets import Dataset
> 
>     # Create argument parser
>     parser = argparse.ArgumentParser(description='Train a sentence transformer model')
>     parser.add_argument('--batch_size', type=int, default=1,
>                         help='Batch size per device (default: 1)')
>     args = parser.parse_args()
> 
>     training_args = SentenceTransformerTrainingArguments(
>         num_train_epochs=3,
>         per_device_train_batch_size=args.batch_size,  # Use the parsed batch size
>         save_strategy='no'
>     )
>     model = SentenceTransformer("microsoft/mpnet-base")
>     train_dataset = Dataset.from_dict({
>         "sentence1": ["It's nice weather outside today.", "He drove to work."],
>         "sentence2": ["It's so sunny.", "She walked to the store."],
>         "score": [1.0, 0.3],
>     })
>     loss = losses.CosineSimilarityLoss(model)
> 
>     trainer = SentenceTransformerTrainer(
>         model=model,
>         train_dataset=train_dataset,
>         loss=loss,
>         args=training_args
>     )
>     trainer.train()
> 
> if __name__ == "__main__":
>     main()
> Agregué esta impresión de depuración **dentro del bucle Trainer antes de accelerator.backward** , que está en la función training_step:
> 
> rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else 0
> for name, param in self.model.named_parameters():
>     if param.grad is not None:
>         print(
>             f"Rank: {rank} | Param: {name} | Device: {param.device} | "
>             f"Grad norm: {param.grad.norm().item()} | Grad:{param.grad} | param :{param} \n"
>         )
>     else:
>         print(
>             f"Rank: {rank} | Param: {name} | Device: {param.device} | Grad: None \n"
>         )
>     break  # Just print the first param for illustration
> **Salida observada**
> 
> ✅ **GPU única:**
> 
> ```
> Rank: 0 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:0 | Grad norm: 0.429... | Grad: [[ 0.0011, 0.0080, ...]]
> ```
> 
> ✅ **DDP (2 GPU):**
> 
> ```
> Rank: 0 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:0 | Grad norm: 0.476... | Grad: [[-0.0053, -0.0001, ...]]
> Rank: 1 | Param: 0.auto_model.embeddings.word_embeddings.weight | Device: cuda:1 | Grad norm: 0.476... | Grad: [[-0.0053, -0.0001, ...]]
> ```
> 
> Como puedes ver, los gradientes son diferentes: no solo están escalados, sino que también tienen signos y valores diferentes.
> 
> **Comportamiento esperado** Los gradientes deben coincidir en ejecuciones de una sola GPU y DDP (hasta la tolerancia numérica) si la configuración, los datos y la pérdida son los mismos.
> 
> Para ejecutar el código:
> 
> GPU única: exportar CUDA_VISIBLE_DEVICES=0 python debug.py --per_device_train_batch_size 2
> 
> 2 GPU: torchrun --nproc_per_node=2 debug.py --per_device_train_batch_size 1
